molgenis-ontology-core/src/main/java/org/molgenis/ontology/core/repository/OntologyTermRepository.java

```diff
@@ -126,217 +148,433 @@ private boolean isOntologyTermExactMatch(Set<String> terms, OntologyTerm ontolog
+		if (ontologyTerm1.getNodePaths().isEmpty() || ontologyTerm2.getNodePaths().isEmpty()) return 0;
 
-		for (Entity ontologyTermNodePathEntity : ontologyTermNodePathEntities)
-		{
-			return ontologyTermNodePathEntity.getString(OntologyTermNodePathMetaData.NODE_PATH);
-		}
-
-		return null;
+		return ontologyTerm1.getNodePaths().stream().flatMap(
+				nodePath1 -> ontologyTerm2.getNodePaths().stream()
+						.map(nodePath2 -> calculateRelatedness(nodePath1, nodePath2)))
+				.max(naturalOrder()).orElse(0.0);
 	}
 
 	/**
 	 * Calculate the distance between nodePaths, e.g. 0[0].1[1].2[2], 0[0].2[1].2[2]. The distance is the non-overlap
 	 * part of the strings
 	 *
-	 * @param nodePath1
-	 * @param nodePath2
-	 * @return distance
+	 * @param nodePath1 the first node path to compare
+	 * @param nodePath2 the second node path to compare
+	 * @return distance the distance between the two node paths.
 	 */
 	public int calculateNodePathDistance(String nodePath1, String nodePath2)
 	{
-		String[] nodePathFragment1 = nodePath1.split("\\.");
-		String[] nodePathFragment2 = nodePath2.split("\\.");
+		String[] nodePathFragment1 = isBlank(nodePath1) ? new String[0] : nodePath1.split(ESCAPED_NODEPATH_SEPARATOR);
+		String[] nodePathFragment2 = isBlank(nodePath2) ? new String[0] : nodePath2.split(ESCAPED_NODEPATH_SEPARATOR);
+
+		int overlapBlock = calculateOverlapBlock(nodePathFragment1, nodePathFragment2);
+		return nodePathFragment1.length + nodePathFragment2.length - overlapBlock * 2;
+	}
+
+	double calculateRelatedness(String nodePath1, String nodePath2)
+	{
+		String[] nodePathFragment1 = isBlank(nodePath1) ? new String[0] : nodePath1.split(ESCAPED_NODEPATH_SEPARATOR);
+		String[] nodePathFragment2 = isBlank(nodePath2) ? new String[0] : nodePath2.split(ESCAPED_NODEPATH_SEPARATOR);
+
+		int overlapBlock = calculateOverlapBlock(nodePathFragment1, nodePathFragment2);
+
+		if (nodePathFragment1.length == 0 || nodePathFragment2.length == 0) return 0;
 
+		return (double) 2 * overlapBlock / (nodePathFragment1.length + nodePathFragment2.length);
+	}
+
+	int calculateOverlapBlock(String[] nodePathFragment1, String[] nodePathFragment2)
+	{
 		int overlapBlock = 0;
 		while (overlapBlock < nodePathFragment1.length && overlapBlock < nodePathFragment2.length
 				&& nodePathFragment1[overlapBlock].equals(nodePathFragment2[overlapBlock]))
 		{
 			overlapBlock++;
 		}
+		return overlapBlock;
+	}
 
-		return nodePathFragment1.length + nodePathFragment2.length - overlapBlock * 2;
+	public Iterable<OntologyTerm> getParents(OntologyTerm ontologyTerm, int maxLevel)
+	{
+		List<OntologyTerm> parentOntologyTerms = new ArrayList<>();
+
+		List<String> nodePaths = ontologyTerm.getNodePaths();
+
+		for (int i = 0; i < maxLevel; i++)
+		{
+			nodePaths = nodePaths.stream().map(this::getParentNodePath).filter(StringUtils::isNotBlank)
+					.collect(toList());
+
+			if (nodePaths.size() > 0)
+			{
+				// TODO: my teammates will take look
```

Comment:
Explanation from Chao:
Node paths are large and therefore stored as text.
Searching for node paths breaks because the separator `.` is a token that gets chopped up by ElasticSearch.

Consider
* Using unique identifiers for the OntologyTerms in the NodePath instead of the 0[1] stuff, so that they're unique
* create a bigger AND AND AND query for each parent node, then the tokenizing shouldn't matter
* Fixing this in a separate story